tools/fp20compiler/main.cpp

```diff
@@ -34,27 +34,68 @@ nvparse_errors errors;
 int line_number;
 char * myin = 0;
 
+static char* find_at_line_start(char* haystack, char* cursor, const char* needle) {
+    cursor = strstr(cursor, needle);
+    while (cursor != NULL) {
+
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return cursor;
+        }
+
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
+
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
+    // Look for the first shader magic
+    char* shader_start = find_at_line_start(ns, ns, "!!");
 
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+    // Warn the user if we couldn't find any shader at all
+    if (shader_start == NULL) {
+        fprintf(stderr, "no shaders found\n");
+    }
 
-        ts10_init(ts_start);
-        ts10_parse();
+    // Loop until we can't find a shader anymore
+    while (shader_start != NULL) {
 
-        *ts_end = '/';
-    }
+        // Find end of shader magic line
+        char* line_end = strchr(shader_start, '\n');
+
+        // The next shader magic marks the end of the current shader; find it
+        char* next_shader_start;
+        if (line_end == NULL) {
+            next_shader_start = NULL;
+        } else {
+            next_shader_start = find_at_line_start(ns, line_end + 1, "!!");
+        }
+
+        // If another shader exists, modify it to end current shader string
+        if (next_shader_start) { *next_shader_start = '\0'; }
+
+        // Process shader
+        if (is_ts10(shader_start)) {
+            ts10_init(shader_start);
+            ts10_parse();
+        } else if (is_rc10(shader_start)) {
+            rc10_init(shader_start);
+            rc10_parse();
+        } else {
+            if (line_end) { *line_end = '\0'; }
+            fprintf(stderr, "unknown shader type \"%s\"\n", shader_start);
```

Comment:
This won't display a meaningful message if `\r` is included (such as windows-line endings).

I consider this a minor issue. We can create an issue after merge.
I finally want this change merged so I can continue with actual work in fp20compiler and XQEMU.